packages/react-storage/src/components/StorageBrowser/views/hooks/usePaginate.ts

```diff
@@ -2,54 +2,59 @@ import React from 'react';
 
 interface UsePaginate {
   currentPage: number;
-  handlePaginateNext: (input: {
-    resultCount: number;
-    hasNextToken: boolean;
-  }) => void;
-  handlePaginatePrevious: (input?: {}) => void;
+  highestPageVisited: number;
+  onPaginate: (page: number) => void;
   handleReset: () => void;
   range: [start: number, end: number];
 }
 
 export const usePaginate = ({
-  onPaginateNext,
-  onPaginatePrevious,
+  hasNextToken,
+  paginateCallback,
   pageSize,
+  resultCount,
 }: {
-  onPaginateNext?: () => void;
-  onPaginatePrevious?: () => void;
+  hasNextToken: boolean;
+  paginateCallback?: () => void;
   pageSize: number;
+  resultCount: number;
 }): UsePaginate => {
   const [currentPage, setCurrentPage] = React.useState(1);
 
   const handleReset = React.useRef(() => {
     setCurrentPage(1);
   }).current;
 
-  return React.useMemo(() => {
+  return React.useMemo((): UsePaginate => {
+    const highestPageVisited = Math.ceil(resultCount / pageSize);
     const isFirstPage = currentPage === 1;
     const start = isFirstPage ? 0 : (currentPage - 1) * pageSize;
     const end = isFirstPage ? pageSize : currentPage * pageSize;
+
     return {
       currentPage,
-      handlePaginateNext: (input) => {
-        const { hasNextToken, resultCount } = input;
-        const highestPageVisited = Math.round(resultCount / pageSize);
+      onPaginate: (page) => {
         const shouldPaginate =
-          highestPageVisited === currentPage && hasNextToken;
-
-        if (shouldPaginate && typeof onPaginateNext === 'function') {
-          onPaginateNext();
+          page >= 1 && (page <= highestPageVisited || hasNextToken);
+        if (shouldPaginate) {
+          if (typeof paginateCallback === 'function') paginateCallback();
+          setCurrentPage(page);
+        } else {
+          // eslint-disable-next-line no-console
+          console.warn('Page is out of bounds');
```

Comment:
Because the responsibility of disabling pagination buttons is on the component rather than provided by the props it receives, this would occur if a developer does not prevent an invalid pagination action from a user. 

This specific approach was suggested in a prior discussion, although I also am unsure if it is the best approach. What would you suggest?